File: reactnative/src/api/AgentStarter.ts
import { KeyCenter } from '../utils/KeyCenter';

// Type declarations for global objects available in React Native
declare const btoa: ((data: string) => string) | undefined;
declare const Buffer: {
  from(data: string, encoding: 'utf-8'): {
    toString(encoding: 'base64'): string;
  };
} | undefined;

interface StartAgentRequest {
  channelName: string;
  agentRtcUid?: string; // 默认 "1009527"
  token: string;
}

interface StartAgentResponse {
  agent_id: string;
}

export class AgentStarter {
  private static readonly API_BASE_URL =
    'https://api.sd-rtn.com/cn/api/conversational-ai-agent/v2/projects';
  private static readonly DEFAULT_AGENT_RTC_UID = '1009527';

  static async startAgentAsync(
    request: StartAgentRequest
  ): Promise<string> {
    const projectId = KeyCenter.AGORA_APP_ID;
    const url = `${this.API_BASE_URL}/${projectId}/join/`;

    // Build request body for RTC+DataStream mode
    // Reference: harmonyos/entry/src/main/ets/api/AgentStarter.ets:113-133
    const properties: Record<string, any> = {
      channel: request.channelName,
      agent_rtc_uid: request.agentRtcUid || this.DEFAULT_AGENT_RTC_UID,
      remote_rtc_uids: ['*'],
      token: request.token,
    };

    // Add parameters for DataStream mode
    const parameters: Record<string, any> = {
      data_channel: 'datastream',
      transcript: {
        enable_words: false,
      },
    };
    properties.parameters = parameters;

    // Add advanced_features to disable RTM (using DataStream instead)
    const advancedFeatures: Record<string, any> = {
      enable_rtm: false,
    };
    properties.advanced_features = advancedFeatures;

    const requestBody = {
      name: request.channelName,
      pipeline_id: KeyCenter.PIPELINE_ID,
      properties: properties,
    };

    const authorization = this.generateAuthorizationHeader(
      KeyCenter.REST_KEY,
      KeyCenter.REST_SECRET
    );

    // Log request details for debugging
    console.log('[AgentStarter] Request URL:', url);
    console.log('[AgentStarter] Request body:', JSON.stringify({
      ...requestBody,
      properties: {
        ...requestBody.properties,
        token: requestBody.properties.token ? '***' : '', // Hide sensitive data
        // Keep parameters and advanced_features visible for debugging
      },
    }, null, 2));

    try {
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: authorization,
        },
        body: JSON.stringify(requestBody),
      });

      console.log('[AgentStarter] Response status:', response.status, response.statusText);

      if (!response.ok) {
        const errorBody = await response.text();
        const errorMessage = `Start agent error: httpCode=${response.status}, httpMsg=${errorBody}`;
        console.error('[AgentStarter]', errorMessage);
        throw new Error(errorMessage);
      }

      const data: StartAgentResponse = await response.json();
      console.log('[AgentStarter] Response data:', JSON.stringify({
        hasAgentId: !!data.agent_id,
        agentId: data.agent_id || 'missing',
      }));

      if (!data.agent_id) {
        const errorMessage = `Failed to parse agentId from response: ${JSON.stringify(data)}`;
        console.error('[AgentStarter]', errorMessage);
        throw new Error(errorMessage);
      }

      return data.agent_id;
    } catch (error: any) {
      // Handle network errors
      if (error instanceof TypeError && error.message.includes('fetch')) {
        const errorMessage = `Start agent network error: ${error.message}`;
        console.error('[AgentStarter]', errorMessage);
        throw new Error(errorMessage);
      }
      // Re-throw other errors
      throw error;
    }
  }

  static async stopAgentAsync(
    projectId: string,
    agentId: string
  ): Promise<void> {
    const url = `${this.API_BASE_URL}/${projectId}/agents/${agentId}/leave`;

    const authorization = this.generateAuthorizationHeader(
      KeyCenter.REST_KEY,
      KeyCenter.REST_SECRET
    );

    // Log request details for debugging
    console.log('[AgentStarter] Stop agent request URL:', url);
    console.log('[AgentStarter] Stop agent request: projectId=', projectId, 'agentId=', agentId);

    try {
      const response = await fetch(url, {
        method: 'POST',
        headers: {
          Authorization: authorization,
        },
      });

      console.log('[AgentStarter] Stop agent response status:', response.status, response.statusText);

      if (!response.ok) {
        const errorBody = await response.text();
        const errorMessage = `Stop agent error: httpCode=${response.status}, httpMsg=${errorBody}`;
        console.error('[AgentStarter]', errorMessage);
        throw new Error(errorMessage);
      }

      console.log('[AgentStarter] Stop agent successfully');
    } catch (error: any) {
      // Handle network errors
      if (error instanceof TypeError && error.message.includes('fetch')) {
        const errorMessage = `Stop agent network error: ${error.message}`;
        console.error('[AgentStarter]', errorMessage);
        throw new Error(errorMessage);
      }
      // Re-throw other errors
      throw error;
    }
  }

  private static generateAuthorizationHeader(
    key: string,
    secret: string
  ): string {
    const credentials = `${key}:${secret}`;
    
    // Use btoa if available, otherwise use Buffer (React Native polyfill)
    let base64: string;
    if (typeof btoa !== 'undefined') {
      base64 = btoa(credentials);
    } else if (typeof Buffer !== 'undefined') {
      base64 = (Buffer as any).from(credentials, 'utf-8').toString('base64');
    } else {
      throw new Error('Neither btoa nor Buffer is available for Base64 encoding');
    }
    
    return `Basic ${base64}`;
  }
}

File: flutter/startup/lib/services/agent_starter.dart
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'app_config.dart';

class AgentStarter {
  static const _jsonType = 'application/json; charset=utf-8';

  static const String _base = 'https://api.sd-rtn.com/cn/api/conversational-ai-agent/v2/projects';

  static String _authHeader() {
    final creds = '${AppConfig.restKey}:${AppConfig.restSecret}';
    return 'Basic ${base64Encode(utf8.encode(creds))}';
  }

  static final Set<String> _sensitive = {
    'auth','token','password','cert','secret','appId','app_id','appCertificate','restKey','restSecret'
  };

  static dynamic _mask(dynamic v) {
    if (v is Map) {
      return v.map((k, value) {
        final key = k.toString();
        if (_sensitive.contains(key)) {
          return MapEntry(key, '***');
        }
        return MapEntry(key, _mask(value));
      });
    } else if (v is List) {
      return v.map(_mask).toList();
    }
    return v;
  }

  static String _maskJson(String input) {
    try {
      final obj = jsonDecode(input);
      return jsonEncode(_mask(obj));
    } catch (_) {
      return input;
    }
  }

  static Future<String> startAgent({required String channelName, required String agentRtcUid, required String token}) async {
    final projectId = AppConfig.appId;
    final url = Uri.parse('$_base/$projectId/join/');
    final body = jsonEncode({
      'name': channelName,
      'pipeline_id': AppConfig.pipelineId,
      'properties': {
        'channel': channelName,
        'agent_rtc_uid': agentRtcUid,
        'remote_rtc_uids': ['*'],
        'token': token,
        'parameters': {
          'transcript': {
            'enable_words': false,
          },
        }
      }
    });
    debugPrint('POST $url');
    debugPrint('Headers: ${jsonEncode({'Authorization': '***', 'Content-Type': _jsonType})}');
    debugPrint('Body: ${_maskJson(body)}');
    final resp = await _postWithRedirects(url, headers: {
      'Authorization': _authHeader(),
      'Content-Type': _jsonType,
    }, body: body);
    debugPrint('Resp ${resp.statusCode} ${resp.reasonPhrase}');
    debugPrint(_maskJson(resp.body));
    if (resp.statusCode >= 200 && resp.statusCode < 300) {
      final json = jsonDecode(resp.body);
      final agentId = (json['agent_id'] ?? '').toString();
      if (agentId.isEmpty) {
        throw Exception('agent_id empty');
      }
      return agentId;
    }
    throw Exception('Start agent error: ${resp.statusCode} ${resp.body}');
  }

  static Future<void> stopAgent(String agentId) async {
    final projectId = AppConfig.appId;
    final url = Uri.parse('$_base/$projectId/agents/$agentId/leave');
    debugPrint('POST $url');
    debugPrint('Headers: ${jsonEncode({'Authorization': '***', 'Content-Type': _jsonType})}');
    final resp = await _postWithRedirects(url, headers: {
      'Authorization': _authHeader(),
      'Content-Type': _jsonType,
    }, body: "");
    debugPrint('Resp ${resp.statusCode} ${resp.reasonPhrase}');
    if (resp.statusCode < 200 || resp.statusCode >= 300) {
      throw Exception('Stop agent error: ${resp.statusCode} ${resp.body}');
    }
  }
  static Future<http.Response> _postWithRedirects(Uri url, {required Map<String, String> headers, required String body, int maxRedirects = 3}) async {
    var current = url;
    for (var i = 0; i <= maxRedirects; i++) {
      final resp = await http.post(current, headers: headers, body: body);
      final code = resp.statusCode;
      if (code == 308 || code == 307 || code == 302 || code == 301) {
        final loc = resp.headers['location'] ?? resp.headers['Location'];
        debugPrint('Redirect ($code) to: ${loc ?? "<none>"}');
        if (loc == null || loc.isEmpty) {
          return resp;
        }
        current = Uri.parse(loc);
        continue;
      }
      return resp;
    }
    return await http.post(current, headers: headers, body: body);
  }
}